subsys/net/l2/ethernet/dsa/dsa.c

```diff
@@ -0,0 +1,138 @@
+/** @file
+ * @brief LLDP related functions
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_dsa);
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <net/net_core.h>
+#include <net/ethernet.h>
+#include <net/net_mgmt.h>
+#include <net/dsa.h>
+
+/*
+ * RECEIVE HANDLING CODE - ingress (ETH -> DSA slave ports)
+ */
+
+static int dsa_check_iface(struct net_if *iface)
+{
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return -ENOENT;
+	}
+
+	if (!((net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT) ||
+	    (net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_SLAVE_PORT))) {
+		return -ESRCH;
+	}
+
+	return 0;
+}
+
+
+int dsa_register_recv_callback(struct net_if *iface, net_dsa_recv_cb_t cb)
+{
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctx = net_if_l2_data(iface);
+
+	if (cb) {
+		ctx->cb = cb;
+	}
+
+	return 0;
+}
+
+enum net_verdict net_dsa_recv(struct net_if *iface, struct net_pkt *pkt)
+{
+
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return NET_DROP;
+	}
+
+	ctx = net_if_l2_data(iface);
+
+	if (ctx->cb) {
+		return ctx->cb(iface, pkt);
+	}
+
+	return NET_CONTINUE;
+}
+
+int dsa_enable_port(struct net_if *iface, u8_t port)
+{
+	return 0;
+}
+
+__weak struct net_pkt *dsa_xmit_pkt(struct device *dev, struct net_pkt *pkt)
+{
+	return pkt;
+}
+
+__weak struct net_if *dsa_get_iface(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+	struct net_linkaddr lladst;
+	struct net_if *iface_sw;
+
```

Comment:
Before the lladdr check, you should check that the interface is ethernet one and that the iface supports DSA.
```
	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
		return iface;
	}
```

The DSA check could be something like this
```
	if (!(net_eth_get_hw_capabilities(iface) &
	     (ETHERNET_DSA_SLAVE_PORT | ETHERNET_DSA_MASTER_PORT))) {
		return iface;
	}
```

Reply:
Ok.